org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java

```diff
@@ -0,0 +1,168 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit4.service;
+
+import static org.osgi.test.common.inject.FieldInjector.findAnnotatedNonStaticFields;
+import static org.osgi.test.common.inject.FieldInjector.setField;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.junit.rules.MethodRule;
+import org.junit.runners.model.FrameworkMethod;
+import org.junit.runners.model.Statement;
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.common.service.ServiceAware;
+import org.osgi.test.junit4.context.BundleContextRule;
+
+/**
+ * A JUnit 4 Rule to depend on OSGi services.
+ * <p>
+ * Example: <br>
+ *
+ * <pre>
+ * &#64;Rule
+ * public ServiceRule sr = new ServiceRule();
+ *
+ * &#64;InjectService
+ * Foo foo;
+ *
+ * &#64;Test
+ * public void aTest() {
+ * 	// use foo
+ * }
+ * </pre>
+ */
+public class ServiceRule implements AutoCloseable, MethodRule {
+
+	private final Map<InjectService, ServiceConfiguration<?>>	configurations	= new ConcurrentHashMap<>();
+
+	private final BundleContextRule								bundleContextRule;
+
+	public ServiceRule() {
+		this(new BundleContextRule());
+	}
+
+	public ServiceRule(BundleContextRule bundleContextRule) {
+		this.bundleContextRule = bundleContextRule;
+	}
+
+	public ServiceRule init(Object testInstance) {
+		bundleContextRule.init(testInstance);
+		List<Field> fields = findAnnotatedNonStaticFields(testInstance.getClass(), InjectService.class);
+
+		fields.forEach(field -> {
+			assertValidFieldCandidate(field);
+
+			InjectService injectService = field.getAnnotation(InjectService.class);
+			Class<?> memberType = field.getType();
+			Type genericMemberType = field.getGenericType();
+
+			setField(field, testInstance, resolveReturnValue(memberType, genericMemberType, injectService,
+				bundleContextRule, configurations));
+		});
+
+		return this;
+	}
+
+	@Override
+	public void close() throws Exception {
+		for (Iterator<Entry<InjectService, ServiceConfiguration<?>>> itr = configurations.entrySet()
+			.iterator(); itr.hasNext();) {
+			Entry<InjectService, ServiceConfiguration<?>> entry = itr.next();
+			entry.getValue()
+				.close();
+			itr.remove();
+		}
+		bundleContextRule.close();
```

Comment:
This may be a problem if the `bundleContextRule` object was supplied to the `ServiceRule(BundleContextRule)` constructor. That is, should our close result in closing a supplied BundleContextRule? If we created the BundleContextRule, as is done by the `ServiceRule()` constructor, then we obviously want to close it. 

Perhaps the simple answer is to get rid of the `ServiceRule(BundleContextRule)` constructor. We just always create a new BundleContextRule. I don't see any use of the `ServiceRule(BundleContextRule)` constructor in this PR.

Reply:
that depends on if we consider the consequences of having:
```
@Rule BundleContextRule bcr = new BundleContextRule();
@Rule ServiceRule sr = new ServiceRule();
```
where now you have double injection happening.
as opposed to having:
```
@Rule BundleContextRule bcr = new BundleContextRule();
@Rule ServiceRule sr = new ServiceRule(bcr);
```
where the internal state of BundleContextRule is managed and avoids the double injection

Reply:
Fine. But then `sr.close()` must not close `bcr`.

Reply:
And what if JUnit calls `bcr.close()` before `sr.close()`? Will 'sr.close()` fail because of its dependence on a closed `bcr`?

Reply:
did you look at the code of BCR close? it should be resonably implemented so it doesn't matter who calls it first?

Reply:
it shouldn't matter if bcr is closed I don't think

Reply:
... this is why I had used inheritance...

Reply:
It looks like this will call ServiceTracker.close which will call BundleContext.ungetService for each tracked service. Assuming the real BundleContext is still valid, then no issue.

Reply:
>  this is why I had used inheritance...

That just means you always had your own copy of a BundleContextRule which is what I was saying by removing the ServiceRule(BundleContextRule) constructor. Same result with aggregation instead of inheritance. Inheritance is a problem since it limits future implementation options.

Reply:
A ServiceRule _is not_ a BundleContextRule. A ServiceRule just needs a BundleContext.

Reply:
fixed all this